sycl/source/detail/config.cpp

```diff
@@ -0,0 +1,138 @@
+//==---------------- config.cpp ---------------------------------*- C++-*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl/detail/config.hpp>
+
+#include <cstring>
+#include <fstream>
+#include <iostream>
+#include <limits>
+#include <sstream>
+
+#define STRINGIFY_LINE_HELP(s) #s
+#define STRINGIFY_LINE(s) STRINGIFY_LINE_HELP(s)
+
+namespace cl {
+namespace sycl {
+namespace detail {
+
+#ifdef DISABLE_CONFIG_FROM_ENV
+constexpr bool ConfigFromEnvEnabled = false;
+#else
+constexpr bool ConfigFromEnvEnabled = true;
+#endif // DISABLE_CONFIG_FROM_ENV
+
+#ifdef DISABLE_CONFIG_FROM_CONFIG_FILE
+constexpr bool ConfigFromFileEnabled = false;
+#else
+constexpr bool ConfigFromFileEnabled = true;
+#endif // DISABLE_CONFIG_FROM_CONFIG_FILE
+
+#ifdef DISABLE_CONFIG_FROM_COMPILE_TIME
+constexpr bool ConfigFromCompileDefEnabled = false;
+#else
+constexpr bool ConfigFromCompileDefEnabled = true;
+#endif // DISABLE_CONFIG_FROM_COMPILE_TIME
+
+#ifndef SYCL_CONFIG_NAME
+#define SYCL_CONFIG_NAME sycl_config.cfg
+#endif // SYCL_CONFIG_NAME
+
+#define CONFIG(Name, MaxSize, CompileTimeDef)                                  \
+  const char *SYCLConfigBase<Name>::MValueFromFile = nullptr;                  \
+  char SYCLConfigBase<Name>::MStorage[MaxSize];
+#include <CL/sycl/detail/config.def>
+#undef CONFIG
+
+template <ConfigID Config> const char *SYCLConfig<Config>::getRawValue() {
+  if (ConfigFromEnvEnabled)
+    if (const char *ValStr = getenv(BaseT::MConfigName))
+      return ValStr;
+
+  if (ConfigFromFileEnabled) {
+    readConfig();
```

Comment:
This seems to execute for every `get` operation, right?

Reply:
Yes, but it has check if it's initialized inside.
It's not atomic thought so races can happen here. But I don't expect this function to be called from multiple threads and even if it's called everything should be OK in terms of functionality.

Reply:
Right, I missed the check inside the readConfig.